pulse2percept/models/granley2021.py

```diff
@@ -0,0 +1,559 @@
+            if hasattr(m, attr):
+                return getattr(self.bright_model, attr)
+        raise AttributeError("%s not found" % attr)
+
+    def __setattr__(self, name, value):
+        """Called when an attribute is set
+
+        This method is called when a new attribute is set(e.g.,
+        ``model.a=2``). This is allowed in the constructor, but will raise a
+        ``FreezeError`` elsewhere.
+
+        ``model.a = X`` can be used as a shorthand to set ``model.bright_model.a``,
+        etc
+        """
+        try:
+            if sys._getframe(2).f_code.co_name == '__init__' or  \
+               sys._getframe(3).f_code.co_name == '__init__':
+                super().__setattr__(name, value)
+                return
+        except FreezeError:
+            pass
+        # Check whether the attribute is a part of any
+        # bright/size/streak model
+        found = False
+        # try to set it ourselves, but can't use get_attr
+        try:
+            self.__getattribute__(name)
+            # if we get here, we have the attribute, not (neccesarily) an effects model
+            super().__setattr__(name, value)
+            found = True
+        except AttributeError:
+            pass
+        for m in [self.bright_model, self.size_model, self.streak_model]:
+            try:
+                if hasattr(m, name):
+                    setattr(m, name, value)
+                    found = True
+            except (AttributeError, FreezeError):
+                pass
+        if not found:
+            err_str = ("'%s' not found. You cannot add attributes to %s "
+                       "outside the constructor." % (name,
+                                                     self.__class__.__name__))
+            raise FreezeError(err_str)
+
+    def get_default_params(self):
+        base_params = super(BiphasicAxonMapSpatial, self).get_default_params()
+        params = {
+            # Callable model used to modulate percept brightness with amplitude,
+            # frequency, and pulse duration
+            'bright_model': None,
+            # Callable model used to modulate percept size with amplitude,
+            # frequency, and pulse duration
+            'size_model': None,
+            # Callable model used to modulate percept streak length with amplitude,
+            # frequency, and pulse duration
+            'streak_model': None,
+            # Use probabilistic thresholding
+            'do_thresholding': False
+        }
+        params.update(base_params)
+        return params
+
+    def _build(self):
+        if self.bright_model is None:
+            self.bright_model = DefaultBrightModel(
+                do_thresholding=self.do_thresholding)
+        if self.size_model is None:
+            self.size_model = DefaultSizeModel(self.rho)
+        if self.streak_model is None:
+            self.streak_model = DefaultStreakModel(self.axlambda)
+        assert(callable(self.bright_model))
```

Comment:
`assert` might be ignored in pip-packaged code. It's better to do `if not callable(...): raise TypeError("'bright_model' should be blah, not bleh")`